### ISSUE TITLE
New Rule Suggestion: UnnecessaryThis

### ISSUE THREAD

**Issue by mgroth0** [detekt/detekt#6953]
## Expected Behavior of the rule

Disallows `this` to be used as a receiver (when there is no `@`)

```kotlin
// Allowed
fun Int.triple1(): Int {
    return this * 3
}

// Allowed
fun Int.triple2(): Int {
    return 3.times(this)
}

// Disallowed
fun Int.triple3(): Int {
    return this.times(3)
}

// Allowed
fun Int.triple4(): Int {
    return times(3)
}

// Allowed 
context(Int,Double)
fun tripleTheInt(): Int {
    return this@Int.times(3)
}

```

## Context

This falls in line with the other UnneccesaryX rules such as UnnecessaryBackticks, UnnecessaryLet, and UnnecessaryParentheses.  It makes code more readable by removing something redundant.

Similar to but distinct from the suggestion for UnnecessaryFullyQualifiedName (https://github.com/detekt/detekt/issues/1501)

**Comment by cortinico** [comment]
That's an interesting rule. The problem is that there might be ambiguous cases such as a little modification to one of the case you suggested:
```diff
// Disallowed
fun Int.triple3(): Int {
+   fun times(input: Int) = 0
    return times(3)
}
```

which actually makes the `this.` necessary.

**Comment by BraisGabin** [comment]
This one, if added, would lean to the controversial ones. I don't see the problem with `this.times(3)` in this case:

```yaml
fun Int.triple3(): Int {
    return this.times(3)
}
```

It is redundant but it gives more context to the reader, I think. Mainly while reviewing a PR.

(I would always write this as `this * 3` but that's other topic because I assume this rule wouldn't complain about that)

**Comment by BraisGabin** [comment]
@detekt/maintainers We should take a deccission over this proposal. I understand that for some use cases this rule can be useful but in general I find that remove all the redundant `this` can harm the readability of the code. So, for that reason I think that this rule should be developed as a third party rule.

**Comment by TWiStErRob** [comment]
Agreed, I would be voting for an ExplicitThis rule if anything for extension functions.

**Comment by schalkms** [comment]
+1 for third party rule

> detekt/maintainers We should take a deccission over this proposal. I understand that for some use cases this rule can be useful but in general I find that remove all the redundant `this` can harm the readability of the code. So, for that reason I think that this rule should be developed as a third party rule.

**Comment by BraisGabin** [comment]
Closing for the reason above

**Comment by cortinico** [comment]
> +1 for third party rule

Yup same for me
